odl/tomo/geometry/detector.py

```diff
@@ -0,0 +1,459 @@
+# Copyright 2014, 2015 The ODL development group
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Detectors."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from abc import ABCMeta, abstractmethod, abstractproperty
+from future import standard_library
+standard_library.install_aliases()
+from future.utils import with_metaclass
+from builtins import object, super
+
+# External
+import numpy as np
+
+# Internal
+from odl.set.domain import IntervalProd
+from odl.discr.grid import TensorGrid
+
+
+__all__ = ('Detector', 'LineDetector', 'Flat2dDetector',
+           'CircleSectionDetector')
+
+
+class Detector(with_metaclass(ABCMeta, object)):
+
+    """Abstract detector class.
+
+    A detector is described by
+
+    * a dimension parameter,
+    * a set of parameters for surface parametrization,
+    * a function mapping motion and surface parameters to the location
+      of a detector point relative to the reference point,
+    * optionally a surface measure function defined on the surface
+      parametrization parameters only
+    * optionally a sampling grid for the parameters
+    """
+
+    def __init__(self, params, grid=None):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        params : `IntervalProd`
+            The parameter set defining the detector area
+        grid : `TensorGrid`, optional
+            A sampling grid for the parameter set, in which it must be
+            contained
+        """
+        if not isinstance(params, IntervalProd):
+            raise TypeError('parameter set {} is not a an interval product.'
+                            ''.format(params))
+
+        if grid is not None:
+            if not isinstance(grid, TensorGrid):
+                raise TypeError('grid {} is not a `TensorGrid` instance.'
+                                ''.format(grid))
+            if not params.contains_set(grid):
+                raise ValueError('grid {} not contained in parameter set {}.'
+                                 ''.format(grid, params))
+
+        self._params = params
+        self._param_grid = grid
+
+    @abstractproperty
+    def ndim(self):
```

Comment:
I'd prefer if this was not a `@abstractproperty` but that it was instead set by a call to `super().__init__` in subclasses, that way we dont need to write this all over the place and can have the doc (and tests) in one place.